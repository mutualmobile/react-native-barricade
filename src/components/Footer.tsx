import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale } from '../utils';

export type FooterProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
  titleStyle: TextStyle;
};

const Footer = (props: FooterProps): JSX.Element => {
  const { onPress, title, titleStyle } = props;
  const { themeColorStyle } = useThemedColor();

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      <TouchableOpacity onPress={onPress} style={styles.footer}>
        <Text numberOfLines={1} style={[styles.title, titleStyle]}>
          {title}
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 50,
    borderTopWidth: 1,
  },
  footer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: hScale(12),
  },
  title: {
    textAlign: 'center',
    fontSize: hScale(18),
  },
});

export { Footer };
