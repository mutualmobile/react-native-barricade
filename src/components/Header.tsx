import React from 'react';
import {
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';

export type ButtonProps = {
  onPress?: (event: NativeSyntheticEvent<any>) => void;
  title: string;
};

export type HeaderProps = {
  headerLeft?: ButtonProps;
  headerRight?: ButtonProps;
  title: string;
};

const Header = (props: HeaderProps): JSX.Element => {
  const { headerLeft, headerRight, title } = props;
  const { themeColorStyle } = useThemedColor();

  const renderButton = (buttonProps: ButtonProps, style: ViewStyle) => {
    return (
      <TouchableOpacity onPress={buttonProps.onPress} style={style}>
        <Text style={[styles.buttonText, themeColorStyle.primary]}>
          {buttonProps.title}
        </Text>
      </TouchableOpacity>
    );
  };

  const renderHeader = () => {
    return (
      <View style={styles.header}>
        <Text
          numberOfLines={1}
          style={[styles.title, themeColorStyle.textDark]}>
          {title}
        </Text>
        {headerLeft && renderButton(headerLeft, styles.headerLeft)}
        <View style={styles.flex} />
        {headerRight && renderButton(headerRight, styles.headerRight)}
      </View>
    );
  };

  return (
    <View
      style={[
        styles.container,
        themeColorStyle.background,
        themeColorStyle.border,
      ]}>
      {renderHeader()}
    </View>
  );
};

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  container: {
    borderBottomWidth: 1,
  },
  header: {
    marginHorizontal: hScale(12),
    height: vScale(50),
    alignItems: 'center',
    flexDirection: 'row',
  },
  title: {
    position: 'absolute',
    textAlign: 'center',
    fontSize: hScale(20),
    fontWeight: 'bold',
    marginHorizontal: hScale(60),
    left: 0,
    right: 0,
  },
  buttonText: {
    fontSize: hScale(18),
    margin: hScale(8),
    textAlign: 'center',
  },
  headerLeft: {
    marginRight: hScale(10),
  },
  headerRight: {
    marginLeft: hScale(10),
  },
});

export { Header };
