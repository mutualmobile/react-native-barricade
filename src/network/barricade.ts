//@ts-ignore
import * as RNFetch from 'react-native/Libraries/Network/fetch';

import { ObjectUtils, UrlUtils } from '../utils';
import {
  Method,
  PathEvaluationCallback,
  PathEvaluationType,
  Request,
  RequestConfig,
  RequestConfigForLib,
  ResponseData,
  ResponseHandler,
} from './barricade.types';
import { interceptor } from './interceptor';
import { createNativeXMLHttpRequest } from './xml-http-request';

export class Barricade {
  running = false;
  private readonly _originalRequestConfig: RequestConfigForLib[] = [];
  private _requestConfig: RequestConfigForLib[] = [];
  private _nativeXMLHttpRequest: typeof XMLHttpRequest;
  private _nativeFetch: RNFetch.fetch;
  private _nativeHeaders: RNFetch.Headers;
  private _nativeRequest: RNFetch.Request;
  private _nativeResponse: RNFetch.Response;

  constructor(requestConfig: RequestConfig[]) {
    this.initRequestConfig(requestConfig);
    this._originalRequestConfig = ObjectUtils.cloneDeep(this.requestConfig);
    this._nativeXMLHttpRequest = global.XMLHttpRequest;
    this._nativeFetch = global.fetch;
    this._nativeHeaders = global.Headers;
    this._nativeRequest = global.Request;
    this._nativeResponse = global.Response;
  }

  set requestConfig(value: RequestConfigForLib[]) {
    this._requestConfig = value;
  }

  get requestConfig() {
    return this._requestConfig;
  }

  handleRequest(request: Request) {
    const method = request._method.toUpperCase() as Method;
    const requestUrl = request._url;
    const parsedRequestUrl = UrlUtils.parseURL(requestUrl);
    request.params = parsedRequestUrl.params;

    const requestConfig = this._requestConfig.find(item => {
      if (
        item.method !== method ||
        !requestUrl.includes(item.pathEvaluation.path)
      ) {
        return;
      } else if (item.pathEvaluation?.type === PathEvaluationType.Includes) {
        return true;
      } else if (item.pathEvaluation?.type === PathEvaluationType.Suffix) {
        return requestUrl.endsWith(item.pathEvaluation.path);
      } else {
        return (item.pathEvaluation as PathEvaluationCallback).callback(
          request,
        );
      }
    });

    if (requestConfig) {
      try {
        const result = (
          requestConfig.responseHandler.find(item => !!item.isSelected) ??
          requestConfig.responseHandler[0]
        ).handler(request);
        if (
          result &&
          typeof (result as PromiseLike<ResponseData>)?.then === 'function'
        ) {
          (result as PromiseLike<ResponseData>).then(resolvedResult => {
            this.resolveRequest(request, resolvedResult, requestConfig.delay);
          });
        } else {
          this.resolveRequest(
            request,
            result as ResponseData,
            requestConfig.delay,
          );
        }
      } catch (error) {
        throw `Barricade intercepted ${requestUrl}(${method}) API and threw an error - ${
          (error as Error).message
        }.`;
      }
    } else {
      createNativeXMLHttpRequest(request, this._nativeXMLHttpRequest);
    }
  }

  resetRequestConfig() {
    this._requestConfig = ObjectUtils.cloneDeep(this._originalRequestConfig);
  }

  private resolveRequest(
    request: Request,
    responseData: ResponseData,
    delay = 400,
  ) {
    let response = responseData.response;
    if (request.responseType === 'blob') {
      // @ts-ignore
      response = (responseData.response as Blob)._data;
    }
    setTimeout(() => {
      request.setResponseData(
        responseData.status,
        responseData.headers,
        response,
      );
    }, delay);
  }

  start() {
    global.XMLHttpRequest = interceptor(this) as any;
    global.fetch = RNFetch.fetch;
    global.Headers = RNFetch.Headers;
    global.Request = RNFetch.Request;
    global.Response = RNFetch.Response;
    this.running = true;
  }

  shutdown() {
    global.XMLHttpRequest = this._nativeXMLHttpRequest;
    global.fetch = this._nativeFetch;
    global.Headers = this._nativeHeaders;
    global.Request = this._nativeRequest;
    global.Response = this._nativeResponse;
    this.running = false;
  }

  initRequestConfig(requests: RequestConfig[]) {
    const updatedRequestConfig = requests.map<RequestConfig>(request => {
      let selectedItem: ResponseHandler | undefined;
      for (let i = 0; i < request.responseHandler.length; i++) {
        if (selectedItem) {
          request.responseHandler[i].isSelected = false;
        } else if (request.responseHandler[i].isSelected) {
          selectedItem = request.responseHandler[i];
        } else if (i === request.responseHandler.length - 1) {
          request.responseHandler[i].isSelected = false;
          request.responseHandler[0].isSelected = true;
          selectedItem = request.responseHandler[0];
        } else {
          request.responseHandler[i].isSelected = false;
        }
      }
      const result = request as RequestConfigForLib;
      result.selectedResponseLabel = selectedItem!.label;

      return result;
    });

    this.requestConfig = updatedRequestConfig;
  }
}
