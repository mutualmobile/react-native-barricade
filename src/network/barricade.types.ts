import { HttpStatusCodeText } from './http-codes';
import { MockedXMLHttpRequest } from './mocked-xml-http-request';

export enum Method {
  Delete = 'DELETE',
  Get = 'GET',
  Head = 'HEAD',
  Options = 'OPTIONS',
  Patch = 'PATCH',
  Post = 'POST',
  Put = 'PUT',
}

export type RequestConfig = {
  path: string;
  method: Method;
  responseHandler: ResponseHandler;
  delay?: number;
};

export type ResponseData = {
  status: keyof typeof HttpStatusCodeText;
  headers: { [k: string]: string };
  response: string;
};

interface ExtraRequestData {
  _url: string;
  _method: string;
  params?: Record<string, string>;
}

export type MockedRequest = MockedXMLHttpRequest & ExtraRequestData;

export type ResponseHandler = {
  (request: MockedRequest): ResponseData | PromiseLike<ResponseData>;
};

export type RequestConfigForMethod = {
  [key in Method]: RequestConfig;
};
export type RequestReferences = Record<string, RequestConfigForMethod>;
