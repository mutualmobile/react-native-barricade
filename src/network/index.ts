export * from './barricade';
export * from './barricade.types';
export * from './http-codes';
