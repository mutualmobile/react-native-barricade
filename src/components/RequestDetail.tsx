import React from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { Strings, Unicode } from '../constants';
import { Barricade, ResponseHandler } from '../network';
import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';
import { Header } from './Header';

type RequestDetailProps = {
  barricade: Barricade;
  selectedListItemIndex?: number;
  onBackPressed: () => void;
};

const RequestDetail = (props: RequestDetailProps): JSX.Element => {
  const { barricade, selectedListItemIndex, onBackPressed } = props;
  const { themeColorStyle } = useThemedColor();

  const onDeailItemPressed = (item: ResponseHandler, index: number) => {
    if (!item.isSelected) {
      const requestConfig = barricade.requestConfig[selectedListItemIndex!];
      requestConfig.responseHandler.map((responseHandler, i) => {
        responseHandler.isSelected = index === i;
        return responseHandler;
      });
      requestConfig.selectedResponseLabel = item.label;
      onBackPressed();
    }
  };

  const renderDetailItem = ({
    item,
    index,
  }: ListRenderItemInfo<ResponseHandler>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onDeailItemPressed(item, index)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        {item.isSelected && (
          <Text style={[styles.icon, themeColorStyle.primary]}>
            {Unicode.CheckMark}
          </Text>
        )}
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{
          title: `${Unicode.ChevronLeft} ${Strings.Back}`,
          onPress: onBackPressed,
        }}
        title={barricade?.requestConfig[selectedListItemIndex!]?.label ?? ''}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={barricade?.requestConfig[selectedListItemIndex!]?.responseHandler}
        renderItem={renderDetailItem}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: hScale(20),
    paddingVertical: vScale(12),
    borderBottomWidth: 1,
  },
  label: {
    fontSize: hScale(18),
    fontWeight: '400',
  },
  icon: {
    fontSize: hScale(18),
  },
});

export { RequestDetail };
