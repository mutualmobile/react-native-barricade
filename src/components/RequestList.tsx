import React, { useState } from 'react';
import {
  FlatList,
  ListRenderItemInfo,
  NativeSyntheticEvent,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';

import { Strings, Unicode } from '../constants';
import { Barricade, RequestConfigForLib } from '../network';
import { useThemedColor } from '../theme';
import { hScale, vScale } from '../utils';
import { Footer } from './Footer';
import { Header } from './Header';

type RequestListProps = {
  barricade: Barricade;
  onDonePressed?: (event: NativeSyntheticEvent<any>) => void;
  onListItemPressed: (index: number) => void;
};

const RequestList = (props: RequestListProps): JSX.Element => {
  const { barricade, onDonePressed, onListItemPressed } = props;
  const { themeColorStyle } = useThemedColor();
  const [refreshList, setRefreshList] = useState(0);

  const onResetPressed = () => {
    barricade?.resetRequestConfig();
    setRefreshList(Math.random());
  };

  const renderListItem = ({
    item,
    index,
  }: ListRenderItemInfo<RequestConfigForLib>) => {
    return (
      <TouchableOpacity
        style={[styles.listItemContainer, themeColorStyle.border]}
        onPress={() => onListItemPressed(index)}>
        <Text style={[styles.label, themeColorStyle.textDark]}>
          {item.label}
        </Text>
        <Text style={[styles.value, themeColorStyle.textLight]}>
          {item.selectedResponseLabel}{' '}
          <Text style={[styles.icon, themeColorStyle.textLight]}>
            {Unicode.ChevronRight}
          </Text>
        </Text>
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.container, themeColorStyle.surface]}>
      <Header
        headerLeft={{ title: Strings.Reset, onPress: onResetPressed }}
        title={Strings.Barricade}
        headerRight={{ title: Strings.Done, onPress: onDonePressed }}
      />
      <FlatList
        style={styles.listContainer}
        contentContainerStyle={[styles.listContainer, themeColorStyle.surface]}
        data={barricade?.requestConfig}
        renderItem={renderListItem}
        extraData={refreshList}
      />
      <Footer barricade={barricade} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  listContainer: {
    flexGrow: 1,
  },
  listItemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: hScale(20),
    paddingVertical: vScale(12),
    borderBottomWidth: 1,
  },
  label: {
    fontSize: hScale(18),
    fontWeight: '400',
  },
  value: {
    fontSize: hScale(16),
  },
  icon: {
    fontSize: hScale(18),
  },
});

export { RequestList };
