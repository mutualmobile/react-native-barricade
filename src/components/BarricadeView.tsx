import React, { useState } from 'react';
import {
  Modal,
  NativeSyntheticEvent,
  SafeAreaView,
  StyleSheet,
} from 'react-native';

import { Barricade } from '../network';
import { ThemeProvider, ThemeType, useThemedColor } from '../theme';
import { RequestDetail } from './RequestDetail';
import { RequestList } from './RequestList';

enum ViewType {
  List,
  Detail,
}

export const BarricadeView = ({
  barricade,
  onRequestClose,
  theme = 'light',
  visible,
}: {
  barricade?: Barricade;
  onRequestClose?: (event: NativeSyntheticEvent<any>) => void;
  theme?: ThemeType;
  visible: boolean;
}) => {
  const { themeColorStyle } = useThemedColor();
  const [viewType, setViewType] = useState<ViewType>(ViewType.List);
  const [selectedListItemIndex, setSelectedListItemIndex] = useState<number>();

  const _onBackPressed = () => {
    setSelectedListItemIndex(undefined);
    setViewType(ViewType.List);
  };

  const _onListItemPressed = (index: number) => {
    setSelectedListItemIndex(index);
    setViewType(ViewType.Detail);
  };

  const renderContent = () => {
    if (viewType === ViewType.List) {
      return (
        <RequestList
          barricade={barricade}
          onListItemPressed={_onListItemPressed}
          onDonePressed={onRequestClose}
        />
      );
    } else {
      return (
        <RequestDetail
          barricade={barricade}
          selectedListItemIndex={selectedListItemIndex}
          onBackPressed={_onBackPressed}
        />
      );
    }
  };

  return (
    <ThemeProvider value={theme}>
      <Modal onRequestClose={onRequestClose} visible={visible}>
        <SafeAreaView
          style={[
            styles.container,
            themeColorStyle.background,
            themeColorStyle.border,
          ]}>
          {renderContent()}
        </SafeAreaView>
      </Modal>
    </ThemeProvider>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
