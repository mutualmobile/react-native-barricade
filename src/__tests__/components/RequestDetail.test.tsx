import React from 'react';
import renderer, { act } from 'react-test-renderer';

import { Footer } from '../../components/Footer';
import { RequestDetail } from '../../components/RequestDetail';
import { Barricade } from '../../network/barricade';
import {
  firstApiConfig,
  secondApiConfig,
  thirdApiConfig,
} from '../data/barricade-test-data';

let barricade: Barricade;
beforeEach(() => {
  barricade = new Barricade([firstApiConfig, secondApiConfig, thirdApiConfig]);
  barricade.start();
});

afterEach(() => {
  jest.clearAllMocks();
  barricade.shutdown();
});

describe('given that API is enabled,', () => {
  test('should match snapshot', () => {
    const selectedIndex = 0;
    barricade.requestConfig[selectedIndex].disabled = false;

    const tree = renderer.create(
      <RequestDetail
        barricade={barricade}
        selectedListItemIndex={selectedIndex}
        onBackPressed={jest.fn}
      />,
    );

    expect(tree).toMatchSnapshot();
    expect.assertions(1);
  });

  test('when "Disable API Mock" button is pressed, should disable API for mocking', async () => {
    const selectedIndex = 0;
    barricade.requestConfig[selectedIndex].disabled = false;

    const tree = renderer.create(
      <RequestDetail
        barricade={barricade}
        selectedListItemIndex={selectedIndex}
        onBackPressed={jest.fn}
      />,
    );
    const instance = tree.root;
    await act(async () => {
      await instance.findByType(Footer).props.onPress();
    });

    expect(barricade.requestConfig[selectedIndex].disabled).toBe(true);
    expect.assertions(1);
  });
});

describe('given that API is disabled,', () => {
  test('should match snapshot', () => {
    const selectedIndex = 0;
    barricade.requestConfig[selectedIndex].disabled = true;

    const tree = renderer.create(
      <RequestDetail
        barricade={barricade}
        selectedListItemIndex={selectedIndex}
        onBackPressed={jest.fn}
      />,
    );

    expect(tree).toMatchSnapshot();
    expect.assertions(1);
  });

  test('when "Enable API Mock" button is pressed, should enable API for mocking', async () => {
    const selectedIndex = 0;
    barricade.requestConfig[selectedIndex].disabled = true;

    const tree = renderer.create(
      <RequestDetail
        barricade={barricade}
        selectedListItemIndex={selectedIndex}
        onBackPressed={jest.fn}
      />,
    );
    const instance = tree.root;
    await act(async () => {
      await instance.findByType(Footer).props.onPress();
    });

    expect(barricade.requestConfig[selectedIndex].disabled).toBe(false);
    expect.assertions(1);
  });
});

test('when an selected list item is pressed, should select list item', async () => {
  const tree = renderer.create(
    <RequestDetail
      barricade={barricade}
      selectedListItemIndex={0}
      onBackPressed={jest.fn}
    />,
  );
  const instance = tree.root;
  await act(() => {
    const findAllByTestID = instance.findAll(
      el => el.props.testID === 'responseListItem0',
    );

    findAllByTestID[0].props.onPress();
  });

  expect(barricade.requestConfig[0].responseHandler[0].isSelected).toBe(true);
  expect.assertions(1);
});

test('when an unselected list item is pressed, should select list item', async () => {
  const tree = renderer.create(
    <RequestDetail
      barricade={barricade}
      selectedListItemIndex={0}
      onBackPressed={jest.fn}
    />,
  );
  const instance = tree.root;
  await act(() => {
    const findAllByTestID = instance.findAll(
      el => el.props.testID === 'responseListItem1',
    );

    findAllByTestID[0].props.onPress();
  });

  expect(barricade.requestConfig[0].responseHandler[1].isSelected).toBe(true);
  expect.assertions(1);
});
