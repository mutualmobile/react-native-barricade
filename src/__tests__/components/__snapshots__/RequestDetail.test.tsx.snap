// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match RequestDetail snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "backgroundColor": "#ffffff",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
        },
        Object {
          "backgroundColor": "#EBEBEB",
        },
        Object {
          "borderColor": "#DEDEDE",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 82.14285714285714,
          "marginHorizontal": 24,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 40,
              "fontWeight": "bold",
              "left": 0,
              "marginHorizontal": 120,
              "position": "absolute",
              "right": 0,
              "textAlign": "center",
            },
            Object {
              "color": "#181818",
            },
          ]
        }
      >
        First API
      </Text>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginRight": 20,
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 36,
                "margin": 16,
                "textAlign": "center",
              },
              Object {
                "color": "#007aa3",
              },
            ]
          }
        >
          ‹ Back
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {
          "flexGrow": 1,
        },
        Object {
          "backgroundColor": "#ffffff",
        },
      ]
    }
    data={
      Array [
        Object {
          "handler": [Function],
          "isSelected": true,
          "label": "Success",
        },
        Object {
          "handler": [Function],
          "isSelected": false,
          "label": "Error",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexGrow": 1,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#DEDEDE",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingHorizontal": 40,
              "paddingVertical": 19.714285714285715,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 36,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#181818",
                },
              ]
            }
          >
            Success
          </Text>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 36,
                },
                Object {
                  "color": "#007aa3",
                },
              ]
            }
          >
            ✓
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderBottomWidth": 1,
              "borderColor": "#DEDEDE",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "opacity": 1,
              "paddingHorizontal": 40,
              "paddingVertical": 19.714285714285715,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontSize": 36,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#181818",
                },
              ]
            }
          >
            Error
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
